packages/manager/src/managers/simulator/SimulatorManager.ts

```diff
@@ -0,0 +1,151 @@
+import fetch from "node-fetch";
+
+import { DecodeError } from "../../lib/DecodeError";
+import { assertPluginsInitialized } from "../../lib/assertPluginsInitialized";
+import { castArray } from "../../lib/castArray";
+import { decode } from "../../lib/decode";
+import { decodeHookResult } from "../../lib/decodeHookResult";
+import { functionCodec } from "../../lib/functionCodec";
+import { markdownToHTML } from "../../lib/markdownToHTML";
+
+import { UnexpectedDataError } from "../../errors";
+
+import { BaseManager } from "../BaseManager";
+
+const sliceSimulatorSetupStepCodec = t.intersection([
+	t.type({
+		title: t.string,
+		body: t.string,
+	}),
+	t.partial({
+		description: t.string,
+		validate: functionCodec,
+	}),
+]);
+
+const SliceSimulatorSetupStepValidationMessageCodec = t.type({
+	title: t.string,
+	message: t.string,
+});
+type SliceSimulatorSetupStepValidationMessageCodec = t.TypeOf<
+	typeof SliceSimulatorSetupStepValidationMessageCodec
+>;
+
+export type SimulatorManagerReadSliceSimulatorSetupStep = {
+	title: string;
+	description?: string;
+	body: string;
+	/**
+	 * Determines if the step is completed.
+	 *
+	 * This proeprty is `undefined` if the project's adapter does not provide a
+	 * validation function for the step; we cannot know if the step is complete
+	 * without a validator.
+	 */
+	isComplete: boolean | undefined;
+	validationMessages: SliceSimulatorSetupStepValidationMessageCodec[];
+};
+
+export type SimulatorManagerReadSliceSimulatorSetupStepsReturnType = {
+	steps: SimulatorManagerReadSliceSimulatorSetupStep[];
+	errors: (DecodeError | HookError)[];
+};
+
+export class SimulatorManager extends BaseManager {
+	async getLocalSliceSimulatorURL(): Promise<string | undefined> {
+		const sliceMachineConfig = await this.project.getSliceMachineConfig();
+
+		return sliceMachineConfig.localSliceSimulatorURL;
+	}
+
+	/**
+	 * @throws {@link UnexpectedDataError} Thrown if the project is not configured
+	 *   with a Slice Simulator URL.
+	 */
+	async checkIsLocalSliceSimulatorURLAccessible(): Promise<boolean> {
+		const localSliceSimulatorURL = await this.getLocalSliceSimulatorURL();
+
+		if (!localSliceSimulatorURL) {
+			throw new UnexpectedDataError(
+				"The project has not been configured with a Slice Simulator URL. Add a `localSliceSimulatorURL` property to your project's configuration to fix this error.",
+			);
+		}
+
+		const res = await fetch(localSliceSimulatorURL);
+
+		return res.ok;
+	}
+
+	async readSliceSimulatorSetupSteps(): Promise<SimulatorManagerReadSliceSimulatorSetupStepsReturnType> {
+		assertPluginsInitialized(this.sliceMachinePluginRunner);
+
+		const hookResult = await this.sliceMachinePluginRunner.callHook(
+			"slice-simulator:setup:read",
+			undefined,
+		);
+		const { data, errors } = decodeHookResult(
+			t.array(sliceSimulatorSetupStepCodec),
+			hookResult,
+		);
+
+		const steps = await Promise.all(
+			data[0].map(async (step) => {
```

Comment:
Yes, because it runs through the `decodeHookResult` validator.

`data` is a nested array.

The outer array contains a list of plugin hook results. `slice-simulator:setup:read` is an adapter-only hook, so we only read the first element and ignore all others. The first plugin is always the adapter, which is required to start Slice Machine.

The first element of that array is an array of Slice Simulator set up steps. We know it is an array because we validated it via `decodeHookResult()`. It may be an empty array, but it will always be an array.